gateway/ln-gateway/src/lib.rs

```diff
@@ -533,14 +533,31 @@ impl Gateway {
             lightning_alias,
             lightning_network,
         };
-        self.set_gateway_state(GatewayState::Running { lightning_context })
-            .await;
+        self.set_gateway_state(GatewayState::Running {
+            lightning_context: lightning_context.clone(),
+        })
+        .await;
         info!("Gateway is running");
 
-        // Blocks until the connection to the lightning node breaks or we receive the
-        // shutdown signal
+        // Runs until the connection to the lightning node breaks or we receive the
+        // shutdown signal.
         if handle
-            .cancel_on_shutdown(self.handle_htlc_stream(stream, handle.clone()))
+            .cancel_on_shutdown(async move {
+                loop {
+                    let htlc_request = match stream.next().await {
```

Comment:
For the gateway safe stop feature to work, we still need to check if the gateway is in the running state and break out of this loop if it is not.

Reply:
I assume you refer to safe stopping where the last HTLCs are still handled but no new ones are accepted?

What about the following order of events:
1. Check here that the GW is still running
2. Wait for the next HTLC event from the stream
3. Initiate a safe shutdown
4. HTLC event is received from stream

Wouldn't we still start processing that HTLC? Maybe I'm misunderstanding the architecture/that feature, but I'd expect a `select!` over `stream.next()` and `is_shutting_down_safely`. Afaiu we cannot rely on the built-in shutdown behavior of `cancel_on_shutdown` because we need to stop HTLC processing first, then wait for in-flight HTLCs and finally shut down normally.